broken-site/broken-site-impl/src/main/java/com/duckduckgo/brokensite/impl/BrokenSiteRefreshesInMemoryStore.kt

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package com.duckduckgo.brokensite.impl
+
+import android.net.Uri
+import com.duckduckgo.brokensite.api.DetectedRefreshPattern
+import com.duckduckgo.brokensite.api.RefreshPattern
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import java.time.Duration
+import java.time.LocalDateTime
+import javax.inject.Inject
+
+interface BrokenSiteRefreshesInMemoryStore {
+    fun resetRefreshCount()
+    fun addRefresh(url: Uri, localDateTime: LocalDateTime)
+    fun getRefreshPatterns(currentDateTime: LocalDateTime): Set<DetectedRefreshPattern>
+}
+
+@ContributesBinding(AppScope::class)
+@SingleInstanceIn(AppScope::class)
+class RealBrokenSiteRefreshesInMemoryStore @Inject constructor() : BrokenSiteRefreshesInMemoryStore {
+
+    private var refreshes: LastRefreshedUrl? = null
+
+    override fun resetRefreshCount() {
+        this.refreshes = null
+    }
+
+    override fun addRefresh(
+        url: Uri,
+        localDateTime: LocalDateTime,
+    ) {
+        refreshes.let {
+            refreshes = if (it == null || it.url != url) {
+                LastRefreshedUrl(url, mutableListOf(localDateTime))
+            } else {
+                it.copy(time = it.time.plus(localDateTime))
+            }
+        }
+    }
+
+    override fun getRefreshPatterns(currentDateTime: LocalDateTime): Set<DetectedRefreshPattern> {
+        val currentRefreshes = refreshes ?: return emptySet()
+        val prunedRefreshes = pruneOldRefreshes(currentDateTime, currentRefreshes)
+
+        val twiceRefreshCount = countPattern(
+            refreshEvents = prunedRefreshes.time,
+            pattern = RefreshPattern.TWICE_IN_12_SECONDS,
+            window = TWICE_REFRESH_WINDOW_IN_SECS,
+        )
+        val thriceRefreshCount = countPattern(
+            refreshEvents = prunedRefreshes.time,
+            pattern = RefreshPattern.THRICE_IN_20_SECONDS,
+            window = THRICE_REFRESH_WINDOW_IN_SECS,
+        )
+
+        val detectedPatterns = mutableSetOf<DetectedRefreshPattern>()
+        if (twiceRefreshCount > 0) {
+            detectedPatterns.add(DetectedRefreshPattern(RefreshPattern.TWICE_IN_12_SECONDS, twiceRefreshCount))
+        }
+        if (thriceRefreshCount > 0) {
+            detectedPatterns.add(DetectedRefreshPattern(RefreshPattern.THRICE_IN_20_SECONDS, thriceRefreshCount))
+        }
+        return detectedPatterns
+    }
+
+    private fun pruneOldRefreshes(
+        currentTime: LocalDateTime,
+        refreshes: LastRefreshedUrl,
+    ): LastRefreshedUrl {
+        val cutoffTime = currentTime.minusSeconds(THRICE_REFRESH_WINDOW_IN_SECS)
+        val newTimes = refreshes.time.filter { it.isAfter(cutoffTime) }
+        return refreshes.copy(time = newTimes)
+    }
+
+    private fun countPattern(refreshEvents: List<LocalDateTime>, pattern: RefreshPattern, window: Long): Int {
```

Comment:
I wonder whether we can just remove the pattern after using it. After all, since we fire the pixel and show the CTA right after refreshing 3 times, we shouldn't need to keep several groups